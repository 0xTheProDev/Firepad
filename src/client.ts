import {
  EventEmitter,
  EventListenerType,
  IEvent,
  IEventEmitter,
} from "./emitter";
import { ITextOperation } from "./text-operation";
import * as Utils from "./utils";

export enum ClientEvent {
  ApplyOperation = "apply",
  SendOperation = "send",
}

interface IClientStateMachine {
  /**
   * Tests whether the Client State is Synchronized with Server or not.
   */
  isSynchronized(): boolean;
  /**
   * Tests whether the Client State is Waiting for Acknowledgement with Server or not.
   */
  isAwaitingConfirm(): boolean;
  /**
   * Tests whether the Client State is Waiting for Acknowledgement with Server along with pending Operation or not.
   */
  isAwaitingWithBuffer(): boolean;
}

export interface IClient extends IClientStateMachine, Utils.IDisposable {
  /**
   * Add listener to Client.
   * @param event - Event name.
   * @param listener - Event handler callback.
   */
  on(event: ClientEvent, listener: EventListenerType<ITextOperation>): void;
  /**
   * Remove listener to Client.
   * @param event - Event name.
   * @param listener - Event handler callback.
   */
  off(event: ClientEvent, listener: EventListenerType<ITextOperation>): void;
  /**
   * Send operation to remote users.
   * @param operation - Text Operation from Editor Adapter
   */
  applyClient(operation: ITextOperation): void;
  /**
   * Recieve operation from remote user.
   * @param operation - Text Operation recieved from remote user.
   */
  applyServer(operation: ITextOperation): void;
  /**
   * Handle acknowledgement
   */
  serverAck(): void;
  /**
   * Handle retry
   */
  serverRetry(): void;
  /**
   * Send operation to Database adapter.
   * @param operation - Text Operation at client end.
   */
  sendOperation(operation: ITextOperation): void;
  /**
   * Apply operation to Editor adapter
   * @param operation - Text Operation at Server end.
   */
  applyOperation(operation: ITextOperation): void;
}

interface IClientSynchronizationState extends IClientStateMachine {
  /**
   * Send operation to remote users.
   * @param operation - Text Operation from Editor Adapter
   */
  applyClient(
    client: IClient,
    operation: ITextOperation
  ): IClientSynchronizationState;
  /**
   * Recieve operation from remote user.
   * @param operation - Text Operation recieved from remote user.
   */
  applyServer(
    client: IClient,
    operation: ITextOperation
  ): IClientSynchronizationState;
  /**
   * Handle acknowledgement
   */
  serverAck(client: IClient): IClientSynchronizationState;
  /**
   * Handle retry
   */
  serverRetry(client: IClient): IClientSynchronizationState;
}

/**
 * In the `Synchronized` state, there is no pending operation that the client
 * has sent to the server.
 */
class Synchronized implements IClientSynchronizationState {
  constructor() {}

  isSynchronized(): boolean {
    return true;
  }

  isAwaitingConfirm(): boolean {
    return false;
  }

  isAwaitingWithBuffer(): boolean {
    return false;
  }

  applyClient(
    client: IClient,
    operation: ITextOperation
  ): IClientSynchronizationState {
    // When the user makes an edit, send the operation to the server and
    // switch to the 'AwaitingConfirm' state
    client.sendOperation(operation);
    return new AwaitingConfirm(operation);
  }

  applyServer(
    client: IClient,
    operation: ITextOperation
  ): IClientSynchronizationState {
    // When we receive a new operation from the server, the operation can be
    // simply applied to the current document
    client.applyOperation(operation);
    return this;
  }

  serverAck(_client: IClient): IClientSynchronizationState {
    Utils.shouldNotGetCalled("There is no pending operation.");
    return this;
  }

  serverRetry(_client: IClient): IClientSynchronizationState {
    Utils.shouldNotGetCalled("There is no pending operation.");
    return this;
  }
}

// Singleton
const _synchronized = new Synchronized();

/**
 * In the `AwaitingConfirm` state, there's one operation the client has sent
 * to the server and is still waiting for an acknowledgement.
 */
class AwaitingConfirm implements IClientSynchronizationState {
  protected readonly _outstanding: ITextOperation;

  constructor(outstanding: ITextOperation) {
    // Save the pending operation
    this._outstanding = outstanding;
  }

  isSynchronized(): boolean {
    return false;
  }

  isAwaitingConfirm(): boolean {
    return true;
  }

  isAwaitingWithBuffer(): boolean {
    return false;
  }

  applyClient(
    _client: IClient,
    operation: ITextOperation
  ): IClientSynchronizationState {
    // When the user makes an edit, don't send the operation immediately,
    // instead switch to 'AwaitingWithBuffer' state
    return new AwaitingWithBuffer(this._outstanding, operation);
  }

  applyServer(
    client: IClient,
    operation: ITextOperation
  ): IClientSynchronizationState {
    // This is another client's operation. Visualization:
    //
    //                   /\
    // this.outstanding /  \ operation
    //                 /    \
    //                 \    /
    //  pair[1]         \  / pair[0] (new outstanding)
    //  (can be applied  \/
    //  to the client's
    //  current document)

    const pair = this._outstanding.transform(operation);
    client.applyOperation(pair[1]);
    return new AwaitingConfirm(pair[0]);
  }

  serverAck(_client: IClient): IClientSynchronizationState {
    // The client's operation has been acknowledged
    // => switch to synchronized state
    return _synchronized;
  }

  serverRetry(client: IClient): IClientSynchronizationState {
    client.sendOperation(this._outstanding);
    return this;
  }
}

/**
 * In the `AwaitingWithBuffer` state, the client is waiting for an operation
 * to be acknowledged by the server while buffering the edits the user makes
 */
class AwaitingWithBuffer implements IClientSynchronizationState {
  protected readonly _outstanding: ITextOperation;
  protected readonly _buffer: ITextOperation;

  constructor(outstanding: ITextOperation, buffer: ITextOperation) {
    // Save the pending operation and the user's edits since then
    this._outstanding = outstanding;
    this._buffer = buffer;
  }

  isSynchronized(): boolean {
    return false;
  }

  isAwaitingConfirm(): boolean {
    return false;
  }

  isAwaitingWithBuffer(): boolean {
    return true;
  }

  applyClient(
    client: IClient,
    operation: ITextOperation
  ): IClientSynchronizationState {
    // Compose the user's changes onto the buffer
    const newBuffer = this._buffer.compose(operation);
    return new AwaitingWithBuffer(this._outstanding, newBuffer);
  }

  applyServer(
    client: IClient,
    operation: ITextOperation
  ): IClientSynchronizationState {
    // Operation comes from another client
    //
    //                       /\
    //     this.outstanding /  \ operation
    //                     /    \
    //                    /\    /
    //       this.buffer /  \* / pair1[0] (new outstanding)
    //                  /    \/
    //                  \    /
    //          pair2[1] \  / pair2[0] (new buffer)
    // the transformed    \/
    // operation -- can
    // be applied to the
    // client's current
    // document
    //
    // * pair1[1]

    const pair1 = this._outstanding.transform(operation);
    const pair2 = this._buffer.transform(pair1[1]);
    client.applyOperation(pair2[1]);
    return new AwaitingWithBuffer(pair1[0], pair2[0]);
  }

  serverAck(client: IClient): IClientSynchronizationState {
    // The pending operation has been acknowledged
    // => send buffer
    client.sendOperation(this._buffer);
    return new AwaitingConfirm(this._buffer);
  }

  serverRetry(client: IClient): IClientSynchronizationState {
    // Merge with our buffer and resend.
    const outstanding = this._outstanding.compose(this._buffer);
    client.sendOperation(outstanding);
    return new AwaitingConfirm(outstanding);
  }
}

export class Client implements IClient {
  protected readonly _emitter: IEventEmitter;

  protected _state: IClientSynchronizationState;

  constructor() {
    this._state = _synchronized;
    this._emitter = new EventEmitter([
      ClientEvent.ApplyOperation,
      ClientEvent.SendOperation,
    ]);
  }

  dispose(): void {
    this._emitter.dispose();
  }

  isSynchronized(): boolean {
    return this._state.isSynchronized();
  }

  isAwaitingConfirm(): boolean {
    return this._state.isAwaitingConfirm();
  }

  isAwaitingWithBuffer(): boolean {
    return this._state.isAwaitingWithBuffer();
  }

  on(event: ClientEvent, listener: EventListenerType<ITextOperation>): void {
    return this._emitter.on(event, listener as EventListenerType<IEvent>);
  }

  off(event: ClientEvent, listener: EventListenerType<ITextOperation>): void {
    return this._emitter.off(event, listener as EventListenerType<IEvent>);
  }

  protected _trigger(event: ClientEvent, eventArgs: ITextOperation): void {
    return this._emitter.trigger(event, eventArgs);
  }

  protected _setState(state: IClientSynchronizationState): void {
    this._state = state;
  }

  applyClient(operation: ITextOperation): void {
    this._setState(this._state.applyClient(this, operation));
  }

  applyServer(operation: ITextOperation): void {
    this._setState(this._state.applyServer(this, operation));
  }

  serverAck(): void {
    this._setState(this._state.serverAck(this));
  }

  serverRetry(): void {
    this._setState(this._state.serverRetry(this));
  }

  sendOperation(operation: ITextOperation): void {
    this._trigger(ClientEvent.SendOperation, operation);
  }

  applyOperation(operation: ITextOperation): void {
    this._trigger(ClientEvent.ApplyOperation, operation);
  }
}
