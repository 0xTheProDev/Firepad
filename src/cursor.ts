import { ITextOp } from "./text-op";
import { ITextOperation } from "./text-operation";

/**
 * JSON Representation of a Cursor Object
 *
 * If `position` and `selectionEnd` are equal that means
 * a single cursor; otherwise, it's a selection between
 * two points.
 */
export type CursorType = {
  /** Starting Position of the Cursor/Selection */
  position: number;
  /** Final Position of the Selection */
  selectionEnd: number;
};

export interface ICursor {
  /**
   * Checks if two Cursor are at same position.
   * @param other - Another Cursor Object (could be null).
   */
  equals(other: ICursor | null): boolean;
  /**
   * Return the more current cursor information.
   * @param other - Another Cursor Object.
   */
  compose(other: ICursor): ICursor;
  /**
   * Update the cursor with respect to an operation.
   * @param operation - Text Operation.
   */
  transform(operation: ITextOperation): ICursor;
  /**
   * Returns JSON representation of the Cursor.
   */
  toJSON(): CursorType;
}

/**
 * A cursor has a `position` and a `selectionEnd`. Both are zero-based indexes
 * into the document. When nothing is selected, `selectionEnd` is equal to
 * `position`. When there is a selection, `position` is always the side of the
 * selection that would move if you pressed an arrow key.
 */
export class Cursor implements ICursor {
  protected readonly _position: number;
  protected readonly _selectionEnd: number;

  /**
   * Creates a Cursor object
   * @param position - Starting position of the Cursor
   * @param selectionEnd - Ending position of the Cursor
   */
  constructor(position: number, selectionEnd: number) {
    this._position = position;
    this._selectionEnd = selectionEnd;
  }

  /**
   * Converts a JSON representation of Cursor into Cursor Object
   */
  static fromJSON({ position, selectionEnd }: CursorType): Cursor {
    return new Cursor(position, selectionEnd);
  }

  equals(other: Cursor | null): boolean {
    if (other == null) {
      return false;
    }

    return (
      this._position === other._position &&
      this._selectionEnd === other._selectionEnd
    );
  }

  compose(other: Cursor): Cursor {
    return other;
  }

  protected _transformIndex(ops: ITextOp[], index: number) {
    let newIndex = index;

    for (const op of ops) {
      if (op.isRetain()) {
        index -= op.chars!;
      } else if (op.isInsert()) {
        newIndex += op.text!.length;
      } else {
        newIndex -= Math.min(index, op.chars!);
        index -= op.chars!;
      }

      if (index < 0) {
        break;
      }
    }

    return newIndex;
  }

  transform(operation: ITextOperation): Cursor {
    const ops: ITextOp[] = operation.getOps();
    const newPosition: number = this._transformIndex(ops, this._position);

    if (this._position === this._selectionEnd) {
      return new Cursor(newPosition, newPosition);
    }

    return new Cursor(
      newPosition,
      this._transformIndex(ops, this._selectionEnd)
    );
  }

  toJSON(): CursorType {
    return {
      position: this._position,
      selectionEnd: this._selectionEnd,
    };
  }
}
