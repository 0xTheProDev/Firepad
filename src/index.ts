export * from "./cursor";
export * from "./database-adapter";
export * from "./editor-adapter";
export * from "./firebase-adapter";
export * from "./firepad";
export * from "./firepad-monaco";
export * from "./monaco-adapter";
export * from "./text-operation";

export { Firepad as default } from "./firepad";
