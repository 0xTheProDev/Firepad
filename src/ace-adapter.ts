import { ICursor, Cursor } from "./cursor";
import {
  ClientIDType,
  EditorAdapterEvent,
  EditorEventCallbackType,
  IEditorAdapter,
  IEditorAdapterEvent,
} from "./editor-adapter";
import { IEventEmitter, EventEmitter, EventListenerType } from "./emitter";
import { ITextOperation, TextOperation } from "./text-operation";
import * as Utils from "./utils";

interface IAceDocument extends AceAjax.Document {
  $lines: string[];
}

interface IAceEditorChangeEventData {
  range: {
    start: AceAjax.Position;
    end: AceAjax.Position;
  };
  action: string; // insert, remove
  lines: string[];
  text: string;
}

interface IAceEditorChangeEvent {
  data: IAceEditorChangeEventData;
}

interface IRemoteCursor extends AceAjax.Range {
  id?: number;
}

export class AceAdapter implements IEditorAdapter {
  protected readonly _ace: AceAjax.Editor;
  protected readonly _aceRange: typeof AceAjax.Range;
  protected readonly _classNames: string[];
  protected readonly _aceDocument: IAceDocument;
  protected readonly _aceSession: AceAjax.IEditSession;
  protected readonly _remoteCursors: Map<ClientIDType, IRemoteCursor>;

  protected _initiated: boolean;
  protected _ignoreChanges: boolean;
  protected _lastDocLines: string[];
  protected _lastCursorRange: AceAjax.Range | null;
  protected _emitter: IEventEmitter | null;

  /**
   * Wraps an Ace editor in adapter to work with rest of Firepad
   * @param aceInstance - Ace Code Editor instance
   * @param avoidListeners - Whether or not propagate changes from editor (optional, defaults to `True`)
   */
  constructor(aceInstance: AceAjax.Editor, avoidListeners: boolean = true) {
    this._ace = aceInstance;
    this._aceRange = window.ace.require("ace/range").Range;
    this._aceSession = this._ace.getSession();
    this._aceDocument = this._aceSession.getDocument() as IAceDocument;
    this._aceDocument.setNewLineMode("unix");

    this._classNames = [];
    this._initiated = false;
    this._ignoreChanges = false;
    this._remoteCursors = new Map<ClientIDType, IRemoteCursor>();

    this._grabDocumentState();

    if (!avoidListeners) {
      this._init();
    }
  }

  protected _init(): void {
    this._emitter = new EventEmitter([
      EditorAdapterEvent.Blur,
      EditorAdapterEvent.Change,
      EditorAdapterEvent.CursorActivity,
      EditorAdapterEvent.Error,
      EditorAdapterEvent.Focus,
      EditorAdapterEvent.Redo,
      EditorAdapterEvent.Undo,
    ]);

    this._ace.on("blur", this._onBlur);
    this._ace.on("focus", this._onFocus);
    this._ace.on("change", this._onChange);
    this._aceSession.selection.on("changeCursor", this._onCursorActivity);
  }

  dispose(): void {
    this._ace.off("blur", this._onBlur);
    this._ace.off("focus", this._onFocus);
    this._ace.off("change", this._onChange);
    this._aceSession.selection.off("changeCursor", this._onCursorActivity);

    if (this._emitter) {
      this._emitter.dispose();
      this._emitter = null;
    }
  }

  protected _grabDocumentState(): void {
    this._lastDocLines = this._aceDocument.getAllLines();
    this._lastCursorRange = this._aceSession.selection.getRange();
  }

  on(
    event: EditorAdapterEvent,
    listener: EventListenerType<IEditorAdapterEvent>
  ): void {
    return this._emitter?.on(event, listener);
  }

  off(
    event: EditorAdapterEvent,
    listener: EventListenerType<IEditorAdapterEvent>
  ): void {
    return this._emitter?.off(event, listener);
  }

  registerCallbacks(callbacks: EditorEventCallbackType): void {
    Object.entries(callbacks).forEach(([event, listener]) => {
      this.on(
        event as EditorAdapterEvent,
        listener as EventListenerType<IEditorAdapterEvent>
      );
    });
  }

  protected _trigger(
    event: EditorAdapterEvent,
    eventArgs: IEditorAdapterEvent | void,
    ...extraArgs: unknown[]
  ): void {
    return this._emitter?.trigger(event, eventArgs || {}, ...extraArgs);
  }

  registerUndo(callback: Utils.VoidFunctionType): void {
    this._ace.undo = callback;
  }

  registerRedo(callback: Utils.VoidFunctionType): void {
    this._ace.redo = callback;
  }

  getCursor(): ICursor {
    let start: number, end: number;

    try {
      start = this._aceDocument.positionToIndex(
        this._aceSession.selection.getRange().start
      );
      end = this._aceDocument.positionToIndex(
        this._aceSession.selection.getRange().end
      );
    } catch (error) {
      console.error(error);

      try {
        // If the new range doesn't work (sometimes with setValue), we'll use the old range
        start = this._aceDocument.positionToIndex(this._lastCursorRange.start);
        end = this._aceDocument.positionToIndex(this._lastCursorRange.end);
      } catch (e2) {
        console.log(
          "Couldn't figure out the cursor range:",
          e2,
          "-- setting it to 0:0."
        );

        [start, end] = [0, 0];
      }
    }

    if (start > end) {
      [start, end] = [end, start];
    }

    return new Cursor(start, end);
  }

  setCursor(cursor: ICursor): void {
    const { position, selectionEnd } = cursor.toJSON();

    let start = this._aceDocument.indexToPosition(position, 0);
    let end = this._aceDocument.indexToPosition(selectionEnd, 0);

    /** If selection is inversed */
    if (position > selectionEnd) {
      [start, end] = [end, start];
    }

    /** Create Selection in the Editor */
    this._aceSession.selection.setSelectionRange(
      // @ts-ignore
      new this._aceRange(start.row, start.column, end.row, end.column)
    );
  }

  setOtherCursor(
    clientID: ClientIDType,
    cursor: ICursor,
    userColor: string,
    userName?: string
  ): Utils.IDisposable {
    let cursorRange = this._remoteCursors.get(clientID);

    if (cursorRange) {
      // @ts-ignore
      cursorRange.start.detach();
      // @ts-ignore
      cursorRange.end.detach();
      this._aceSession.removeMarker(cursorRange.id);
    }

    const { position, selectionEnd } = cursor.toJSON();

    // let start = this._aceDocument.indexToPosition(position, 1);
    // let end = this._aceDocument.indexToPosition(selectionEnd, 1);
    let start = this._posFromIndex(position);
    let end = this._posFromIndex(selectionEnd);

    /** If selection is inversed */
    if (position > selectionEnd) {
      [start, end] = [end, start];
    }

    let selectionColor = userColor;
    let className = `remote-client-selection-${userColor.replace("#", "")}`;

    if (position === selectionEnd) {
      /** It's a single cursor */
      selectionColor = "transparent";
      className = className.replace("selection", "cursor");
    }

    console.log("cursor", position, start);

    /** Generate Style rules and add them to document */
    this._addStyleRule(className, selectionColor, userColor);

    // @ts-ignore
    cursorRange = new this._aceRange(
      start.row,
      start.column,
      end.row,
      end.column
    );

    console.log(cursorRange);
    let originalClipRows = cursorRange.clipRows;
    cursorRange.clipRows = (firstRow: number, lastRow: number) => {
      const range = originalClipRows.call(cursorRange, firstRow, lastRow);
      range.isEmpty = () => false;
      return range;
    };

    // @ts-ignore
    cursorRange.start = this._aceDocument.createAnchor(
      cursorRange.start.row,
      cursorRange.start.column
    );

    // @ts-ignore
    cursorRange.end = this._aceDocument.createAnchor(
      cursorRange.end.row,
      cursorRange.end.column
    );
    cursorRange.id = this._aceSession.addMarker(
      cursorRange,
      className,
      "text",
      false
    );

    return {
      dispose: () => {
        const remoteCursor: IRemoteCursor | void = this._remoteCursors.get(
          clientID
        );

        if (!remoteCursor) {
          // Already disposed, nothing to do.
          return;
        }

        // Dispose marker added.
        // @ts-ignore
        remoteCursor.start.detach();
        // @ts-ignore
        remoteCursor.end.detach();
        this._aceSession.removeMarker(remoteCursor.id);
      },
    };
  }

  getText(): string {
    return this._aceDocument.getValue();
  }

  setText(text: string): void {
    this._aceDocument.setValue(text);
  }

  setInitiated(init: boolean): void {
    // Perfomance boost on clearing editor after network calls (do not directly setValue or EOL will get reset and break sync)
    this.setText("");
    this._initiated = init;
  }

  protected _posFromIndex(index: number): AceAjax.Position {
    var j, len, line, ref, row;
    ref = this._aceDocument.$lines;
    for (row = j = 0, len = ref.length; j < len; row = ++j) {
      line = ref[row];
      if (index <= line.length) {
        break;
      }
      index -= line.length + 1;
    }
    return {
      row: row + 1,
      column: index,
    };
  }

  protected _indexFromPosfunction(
    position: AceAjax.Position,
    lines?: string[]
  ) {
    var i, index, j, ref;
    lines ||= this._lastDocLines;
    index = 0;
    for (
      i = j = 0, ref = position.row;
      0 <= ref ? j < ref : j > ref;
      i = 0 <= ref ? ++j : --j
    ) {
      index += this._lastDocLines[i].length + 1;
    }
    return (index += position.column);
  }

  /**
   * Transforms changes from Ace editor into Edit Operations.
   * @param changes - Change Event with List of changes.
   */
  protected _operationFromACEChange(
    change: AceAjax.EditorChangeEvent | IAceEditorChangeEvent
  ): [ITextOperation, ITextOperation] {
    let action: string, start: number, text: string;

    // @ts-ignore - Support for Ace < 1.2.0
    if (change.data) {
      const delta = (change as IAceEditorChangeEvent).data;

      if (delta.action === "insertLines" || delta.action === "removeLines") {
        text =
          delta.lines.join(Utils.EndOfLineSequence.LF) +
          Utils.EndOfLineSequence.LF;
        action = delta.action.replace("Lines", "");
      } else {
        text = delta.text.replace(
          this._aceDocument.getNewLineCharacter(),
          Utils.EndOfLineSequence.LF
        );
        action = delta.action.replace("Text", "");
      }
      start = this._aceDocument.positionToIndex(delta.range.start);
    } else {
      // Ace 1.2.0+
      const {
        action: actionName,
        lines,
        start: startPosition,
      } = change as AceAjax.EditorChangeEvent;
      action = actionName;
      text = lines.join(Utils.EndOfLineSequence.LF);
      start = this._aceDocument.positionToIndex(startPosition);
    }

    let restLength =
      this._lastDocLines.join(Utils.EndOfLineSequence.LF).length - start;

    if (action === "remove") {
      restLength -= text.length;
    }

    const insert_op = new TextOperation()
      .retain(start, null)
      .insert(text, null)
      .retain(restLength, null);
    const delete_op = new TextOperation()
      .retain(start, null)
      .delete(text)
      .retain(restLength, null);

    if (action === "remove") {
      return [delete_op, insert_op];
    }

    return [insert_op, delete_op];
  }

  /**
   * Applies Edit Operations into Ace editor.
   * @param operation - Edit Operations.
   */
  protected _applyOperationToACE(operation: ITextOperation) {
    let index = 0;
    const ops = operation.getOps();

    for (const op of ops) {
      if (op.isRetain()) {
        index += op.chars;
        continue;
      }

      if (op.isInsert()) {
        this._aceDocument.insert(
          this._aceDocument.indexToPosition(index, 0),
          op.text
        );
        index += op.text!.length;
        continue;
      }

      if (op.isDelete()) {
        const start = this._aceDocument.indexToPosition(index, 0);
        const end = this._aceDocument.indexToPosition(index + op.chars, 0);
        // @ts-ignore
        const range = this._aceRange.fromPoints(start, end);
        this._aceDocument.remove(range);
      }
    }

    this._grabDocumentState();
  }

  applyOperation(operation: ITextOperation): void {
    if (!operation.isNoop()) {
      this._ignoreChanges = true;
    }

    this._applyOperationToACE(operation);
    this._ignoreChanges = false;
  }

  invertOperation(operation: ITextOperation): ITextOperation {
    return operation.invert(this.getText());
  }

  protected _onBlur = () => {
    if (!this._ace.selection.isEmpty()) {
      return;
    }

    this._trigger(EditorAdapterEvent.Blur);
  };

  protected _onFocus = () => {
    this._trigger(EditorAdapterEvent.Focus);
  };

  protected _onCursorActivity = () => {
    this._trigger(EditorAdapterEvent.CursorActivity);
  };

  protected _onChange = (
    change: AceAjax.EditorChangeEvent | IAceEditorChangeEvent
  ) => {
    /** Ignore if change is being applied by firepad itself. */
    if (this._ignoreChanges || !this._initiated) {
      return;
    }

    const [mainOp, reverseOp] = this._operationFromACEChange(change);
    this._trigger(EditorAdapterEvent.Change, mainOp, reverseOp);

    this._grabDocumentState();
  };

  /**
   * Returns CSS Style rules for Cursor and Selection.
   * @param className - CSS Classname for the Cursor or Selection.
   * @param backgroundColor - Background color for selection, `transparent` for cursor.
   * @param fontColor - Color of cursor.
   * @returns
   */
  protected _getStyles(
    className: string,
    backgroundColor: string,
    fontColor: string
  ): string {
    return `
      .${className} {
        position: relative;
        background-color: ${backgroundColor};
        border-left: 2px solid ${fontColor};
      }
    `;
  }

  /**
   * Adds CSS Style rules into DOM
   * @param className - CSS Classname for the Cursor or Selection.
   * @param backgroundColor - Background color for selection, `transparent` for cursor.
   * @param fontColor - Color of cursor.
   */
  protected _addStyleRule(
    className: string,
    backgroundColor: string,
    fontColor: string
  ): void {
    Utils.validateTruth(document != null, "This package must run on browser!");

    /** Do not re-inject if already exists in DOM */
    if (this._classNames.includes(className)) {
      return;
    }

    const style = this._getStyles(className, backgroundColor, fontColor);
    const styleTextNode = document.createTextNode(style);
    const styleElement = document.createElement("style");
    styleElement.appendChild(styleTextNode);
    document.head.appendChild(styleElement);

    this._classNames.push(className);
  }
}
