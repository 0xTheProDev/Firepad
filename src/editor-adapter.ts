import { ICursor } from "./cursor";
import { EventListenerType, IEvent } from "./emitter";
import { ITextOperation } from "./text-operation";
import { IDisposable, VoidFunctionType } from "./utils";

export type UndoRedoCallbackType = VoidFunctionType;
export type ClientIDType = string | number;

/** Event handler callback Parameters for `change` event */
type EditorOnChangeFunctionType = (
  operation: ITextOperation,
  inverse: ITextOperation
) => void;

/** Additional State Information about Editor Adapter */
export type EditorAdapterStateType = {
  retain: number;
  skippedChars: number;
  contentLength: number;
};

/** Event parameters for `error` event on Editor Adapter */
export type EditorOnErrorFunctionType = (
  /** Original Error event */
  error: Error,
  /** String representation of the Text Operation */
  operation: string,
  /** Additional State Information about Editor Adapter */
  state: EditorAdapterStateType
) => void;

export enum EditorAdapterEvent {
  Error = "error",
  Blur = "blur",
  Focus = "focus",
  Undo = "undo",
  Redo = "redo",
  Change = "change",
  CursorActivity = "cursorActivity",
}

export type EditorEventCallbackType = {
  [EditorAdapterEvent.Blur]: VoidFunctionType;
  [EditorAdapterEvent.Focus]: VoidFunctionType;
  [EditorAdapterEvent.Error]: EditorOnErrorFunctionType;
  [EditorAdapterEvent.Change]: EditorOnChangeFunctionType;
  [EditorAdapterEvent.CursorActivity]: VoidFunctionType;
};

export interface IEditorAdapterEvent extends IEvent {}

export interface IEditorAdapter extends IDisposable {
  /**
   * Add listener to Editor Adapter.
   * @param event - Event name.
   * @param listener - Event handler callback.
   */
  on(
    event: EditorAdapterEvent,
    listener: EventListenerType<IEditorAdapterEvent>
  ): void;
  /**
   * Remove listener to Editor Adapter.
   * @param event - Event name.
   * @param listener - Event handler callback.
   */
  off(
    event: EditorAdapterEvent,
    listener: EventListenerType<IEditorAdapterEvent>
  ): void;
  /**
   * Attaches callback for Editor Event handling from top-level.
   * @param callbacks - Map of Editor Events and Callbacks.
   */
  registerCallbacks(callbacks: EditorEventCallbackType): void;
  /**
   * Attaches callback for Editor Event handling from top-level.
   * @param callback - Undo Event Handler.
   */
  registerUndo(callback: UndoRedoCallbackType): void;
  /**
   * Attaches callback for Editor Event handling from top-level.
   * @param callback - Redo Event Handler.
   */
  registerRedo(callback: UndoRedoCallbackType): void;
  /**
   * Returns Cursor position of current User in Editor.
   */
  getCursor(): ICursor | null;
  /**
   * Add Cursor position of current User in Editor.
   * @param cursor - Cursor position of Current User.
   */
  setCursor(cursor: ICursor): void;
  /**
   * Add Cursor position of Remote Users in Editor.
   * @param clientID - Remote User ID.
   * @param cursor - Cursor Object of Remote User.
   * @param userColor - HexaDecimal/RGB Color Code for Cursor/Selection.
   * @param userName - User Name to show on Cursor (optional).
   */
  setOtherCursor(
    clientID: ClientIDType,
    cursor: ICursor,
    userColor: string,
    userName?: string
  ): IDisposable;
  /**
   * Returns current content of the Editor.
   */
  getText(): string;
  /**
   * Sets current content of the Editor.
   * @param text - Text Content.
   */
  setText(text: string): void;
  /**
   * Applies operation into Editor.
   * @param operation - Text Operation.
   */
  setInitiated(init: boolean): void;
  /**
   * Sets the inititated boolean which in turn allows onChange events to progress
   * @param _initiated - initiated boolean that represent initial firebase Revisions.
   */
  applyOperation(operation: ITextOperation): void;
  /**
   * Returns invert operation based on current Editor content
   * @param operation - Text Operation.
   */
  invertOperation(operation: ITextOperation): ITextOperation;
}
