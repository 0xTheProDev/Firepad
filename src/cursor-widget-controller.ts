import * as monaco from "monaco-editor";

import { CursorWidget, ICursorWidget } from "./cursor-widget";
import { ClientIDType } from "./editor-adapter";
import { IDisposable } from "./utils";

export interface ICursorWidgetController extends IDisposable {
  /**
   * Add Cursor Widget for the first time for new User.
   * @param clientId - Unique Identifier for remote User.
   * @param range - Position of the Widget.
   * @param userColor - Border Color of the Widget.
   * @param userName - User's name to show up in Widget. (optional, default is set to User ID).
   */
  addCursor(
    clientId: ClientIDType,
    range: monaco.Range,
    userColor: string,
    userName?: string
  ): void;
  /**
   * Update Cursor Widget for existing User, or add a new Cursor Widget if new User.
   * @param clientId - Unique Identifier for remote User.
   * @param range - Position of the Widget.
   * @param userColor - Border Color of the Widget.
   * @param userName - User's name to show up in Widget. (optional, default is set to User ID).
   */
  updateCursor(
    clientId: ClientIDType,
    range: monaco.Range,
    userColor: string,
    userName?: string
  ): void;
  /**
   * Dispose Cursor Widget for existing User.
   * @param clientId - Unique Identifier for remote User.
   */
  removeCursor(clientId: ClientIDType): void;
}

export class CursorWidgetController implements ICursorWidgetController {
  protected readonly _cursors: Map<ClientIDType, ICursorWidget>;
  protected readonly _editor: monaco.editor.IStandaloneCodeEditor;
  protected readonly _tooltipDuration: number;

  constructor(editor: monaco.editor.IStandaloneCodeEditor) {
    this._editor = editor;
    this._tooltipDuration = 1000;
    this._cursors = new Map<ClientIDType, ICursorWidget>();
  }

  addCursor(
    clientId: ClientIDType,
    range: monaco.Range,
    userColor: string,
    userName?: string
  ): void {
    const cursorWidget = new CursorWidget({
      codeEditor: this._editor,
      widgetId: clientId,
      color: userColor,
      range,
      label: userName || clientId.toString(),
      tooltipDuration: this._tooltipDuration,
      onDisposed: () => {
        this.removeCursor(clientId);
      },
    });

    this._cursors.set(clientId, cursorWidget);
  }

  removeCursor(clientId: ClientIDType): void {
    const cursorWidget = this._cursors.get(clientId);

    if (!cursorWidget) {
      /** Already disposed, nothing to do here. */
      return;
    }

    cursorWidget.dispose();
    this._cursors.delete(clientId);
  }

  updateCursor(
    clientId: ClientIDType,
    range: monaco.Range,
    userColor: string,
    userName?: string
  ): void {
    const cursorWidget = this._cursors.get(clientId);

    if (cursorWidget) {
      /** Widget already present, simple update should suffice. */
      cursorWidget.updatePosition(range);
      cursorWidget.updateContent(userName);
      return;
    }

    this.addCursor(clientId, range, userColor, userName);
  }

  dispose(): void {
    this._cursors.forEach((cursorWidget: ICursorWidget) =>
      cursorWidget.dispose()
    );
    this._cursors.clear();
  }
}
