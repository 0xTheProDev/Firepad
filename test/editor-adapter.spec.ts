import { EditorAdapterEvent } from "../src/editor-adapter";

describe("Editor Adapter", () => {
  it("should only listen to finite number of events", () => {
    expect(EditorAdapterEvent).toMatchSnapshot();
  });
});
