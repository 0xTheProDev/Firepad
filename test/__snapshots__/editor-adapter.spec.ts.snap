// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Editor Adapter should only listen to finite number of events 1`] = `
Object {
  "Blur": "blur",
  "Change": "change",
  "CursorActivity": "cursorActivity",
  "Error": "error",
  "Focus": "focus",
  "Redo": "redo",
  "Undo": "undo",
}
`;
